### PR TITLE
Created 'MonotonePolygon'

### PR DESCRIPTION
**Pull request by 1ndy**
I'd like to attempt random monotone polygon generation

TODO:
 - [x] Generate points in counter-clockwise order. Going counter-clockwise from a minimum point to a maximum point means the 'right-half' should be in between the extremes and the 'left-half' should be after.
 - [x] Generalize the function to accept a list of points instead of randomly generating them. All in all, the type signatures should look like this:
  ```haskell
randomMonotone :: RandomGen g => Int -> Rand g (SimplePolygon () Rational)
randomMonotoneDirected :: RandomGen g => Int -> Vector 2 Rational -> Rand g (SimplePolygon () Rational)
monotoneFrom :: (Ord r, Num r) => Vector 2 r -> [Point 2 r -> SimplePolygon () r
  ```

`randomMonotone` should generate a random `Vector` and call `randomMonotoneDirected`.
`randomMonotoneDirected` should generate a set of random points and call `randomMonotoneFrom`. Splitting up the functions like this makes it a lot easier to test them.

### CODE REVIEW

**Comment by Lemmih** [comment]
Excellent!

To rehash the algorithm:

1. Create N `Point 2 Rational` (N >= 3)
2. Create a random `Vector 2 Rational`
3. Find the extremes (min and max) of the points when sorted in the direction of the vector.
    We already have code for this. See `maximumBy (cmpExtreme vector)` and `minimumBy (cmpExtreme vector)`.
4. Take out the two extremal points from the set.
5. Partition the remaining points according to whether they're on the left side or right side of the imaginary line between the two extremal points.
6. Sort the two partitioned sets, one in the direction of the vector and one in the opposite direction.
7. Connect the points, starting from the minimal extreme point, going through the set of points that are increasing in the direction of the vector, then to the maximal point, and finally down through the points that are decreasing in the direction of the vector.

I'll add some type-signatures and tests to your branch.

**Comment by Lemmih** [comment]
Could you tick "Allow edits from maintainers" if you haven't already. Here's a guide: https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork

**Comment by 1ndy** [comment]
Box is checked. Thanks for giving me a starting point, I'll get started tomorrow

**Comment by Lemmih** [comment]
For random Rational points, see this code:

```haskell
granularity :: Integer
granularity = 10000000

-- Random point between screenTop/screenBottom.
genPoint :: RandomGen g => Rand g (Point 2 Rational)
genPoint = do
  x <- liftRand $ randomR (0, granularity)
  y <- liftRand $ randomR (0, granularity)
  pure $ Point2
    ((x % granularity) * screenHeight - screenTop)
    ((y % granularity) * screenHeight - screenTop)
```

It has to be in the `Rand` monad and cannot just be a `Point 2 Rational`.

**Comment by Lemmih** [comment]
Also, you can use the `ClosedLineSegment` constructor to make a line. But the line is actually not needed. It can be imaginary. We just need to partition the points according to which side they're on. We can test this by just using the 'min' and 'max' points you already found.

**Comment by Lemmih** [comment]
Your solution is very close to being correct. It just has to be monadic and use `do`. Like this:
```haskell
randomMonotone :: RandomGen g => Int -> Vector 2 Rational -> Rand g (SimplePolygon () Rational)
randomMonotone nVertices direction = do
  -- 1, skip 2 in this function bc `direction` is given
  points <- replicateM nVertices genPoint
  -- 3
  let min = Data.Geometry.Polygon.Core.minimumBy (cmpExtreme direction) points
      max = Data.Geometry.Polygon.Core.maximumBy (cmpExtreme direction) points
  -- rest of your code here.
```

**Comment by Lemmih** [comment]
Another hint: There's a built-in function for testing which side a point is on.
```haskell
toTheLeft :: Point 2 r -> Bool
toTheLeft x = ccw min max x == CCW
```
You can then use `Data.List.partition :: (a -> Bool) -> [a] -> ([a], [a])` to split the points in two.

**Comment by Lemmih** [comment]
This code:
```haskell
        polygon = SimplePolygon $ Data.CircularSeq.fromList (min : leftHalf : max : rightHalf)
```

Can be written as:

```haskell
        polygon = fromPoints ([min] ++ leftHalf ++ [max] ++ rightHalf)
```

Let me know if I'm micro-managing too much and you'd rather want some peace and quiet. :)

**Comment by 1ndy** [comment]
No you're fine, I appreciate the advice. I am pretty new to haskell and
very new to this codebase so it is helpful to see these shortcuts. I've got
some other responsibilities to tend to to wrap up this work week but I'll
get back to this very soon and incorporate your suggestions.

**Comment by 1ndy** [comment]
Sorry for all the commits. I am getting an error while building `hgeometry-ipe` so this is the only way I can check my code right now

**Comment by noinia** [comment]
What error do you get when building hgeometry-ipe? 

As a workaround (i.e. so that locally you can at least properly compile stuff), you can probably also remove hgeometry-ipe from the cabal.project file as well as disable tests. Not really an ideal solution though.

**Comment by 1ndy** [comment]
```
hgeometry-ipe  > Building library for hgeometry-ipe-0.11.0.0..
hgeometry-ipe  > [12 of 22] Compiling Data.Geometry.Ipe.Reader
hgeometry-ipe  > /home/hunter/hgeometry/hgeometry-ipe/src/Data/Geometry/Ipe/Reader.hs:208:46: error:
hgeometry-ipe  >     Not in scope: ‘Polygon.unsafeFromPoints’
hgeometry-ipe  >     Perhaps you meant ‘Polygon.fromPoints’ (imported from Data.Geometry.Polygon)
hgeometry-ipe  >     Module ‘Data.Geometry.Polygon’ does not export ‘unsafeFromPoints’.
hgeometry-ipe  >     |
hgeometry-ipe  > 208 |                                       poly = Polygon.unsafeFromPoints . dropRepeats $ pts
hgeometry-ipe  >     |                                              ^^^^^^^^^^^^^^^^^^^^^^^^
```
I'll try your suggestion

**Comment by noinia** [comment]
I think you may want to pull the latest changes from origin/master into master, and merge the master branch into this branch.

File: hgeometry/src/Algorithms/Geometry/MonotonePolygon/MonotonePolygon.hs [hunk 1]
```diff
@@ -64,42 +66,61 @@ isMonotone direction p = all isMonotoneAt (map _core $ toPoints p)
...
 
 granularity :: Integer
 granularity = 10000000
 
+toTheLeft :: Point 2 Rational :+ () -> Point 2 Rational :+ () -> Point 2 Rational :+ () -> Bool
+toTheLeft min max x = Data.Geometry.Point.Orientation.Degenerate.ccw a b c == CCW
+    where
+        a = _core min
+        b = _core max
+        c = _core x
```
Comment (Lemmih) [inline hgeometry/src/Algorithms/Geometry/MonotonePolygon/MonotonePolygon.hs:112]: The `ccw'` helper function calls `_core` for you.

`ccw' :: (Ord r, Num r) => Point 2 r :+ a -> Point 2 r :+ b -> Point 2 r :+ c -> CCW`

**Comment by Lemmih** [comment]
> Sorry for all the commits. I am getting an error while building `hgeometry-ipe` so this is the only way I can check my code right now

That's strange. Hopefully merging in master fixed that issue. You can also run `cabal build` from the `hgeometry` sub-directory. Then only `hgeometry-combinatorial` and `hgeometry` will be built.

**Comment by Lemmih** [comment]
I will push some visualization code to make debugging easier.

**Comment by Lemmih** [comment]
Pushed.
If you go to the `hgeometry-showcase` folder and run `stack repl hgeometry-showcase`, it'll give you a GHCi prompt. In GHCi run `:main random_monotone` and an animation will appear in a browser window. It only works with stack and not cabal.

**Comment by Lemmih** [comment]
Ah, I made a mistake in `isMonotone` :D Will push a fix.

**Comment by Lemmih** [comment]
I love it! Your code works really well:

https://user-images.githubusercontent.com/1443107/104795849-6ac31400-57ec-11eb-9b5f-20063cceff00.mp4

**Comment by Lemmih** [comment]
I added two TODO items to your original post.

**Comment by 1ndy** [comment]
Hey that's pretty cool! I'll work on those changes, I like the idea of a generalized `randomMonotoneFrom` function, though since it doesn't use the random monad maybe we should just call it `monotoneFrom`? I'll export it as the former for now

**Comment by Lemmih** [comment]
> Hey that's pretty cool! I'll work on those changes, I like the idea of a generalized `randomMonotoneFrom` function, though since it doesn't use the random monad maybe we should just call it `monotoneFrom`? I'll export it as the former for now

Oh yeah, `monotoneFrom` is a much better name.

**Comment by Lemmih** [comment]
I've deleted some of the code I copy&pasted into your module at the start of the PR. It's been replaced by a better and more standard way of generating random points. Your code is unchanged except for a few type signatures.

**Comment by 1ndy** [comment]
Looks good, are we ready to move it to the proper directory and submit the PR?

**Comment by Lemmih** [comment]
Yes, everything looks good. I'll move the module, write a QuickCheck test, and mark the PR as ready.

**Comment by Lemmih** [comment]
Alright, looks like everything is in order:
 - [x] Implementation
 - [x] Tests
 - [x] Documentation
 - [x] Showcase animation: https://hgeometry.org/#random-monotone
 - [x] Copyright notice to @1ndy (don't know your name so I just wrote 1ndy, feel free to change it)

I'll go ahead and merge this PR. Thank you so much, @1ndy. You've improved HGeometry. I hope we'll see more algorithms from you in the future.

**Comment by 1ndy** [comment]
Thanks for all your help!!

**Pull request merged**